cupyx/scipy/signal/_fir_filter_design.py

```diff
@@ -0,0 +1,168 @@
+    if (cupy.diff(bands) <= 0).any() or (cupy.diff(bands[:, 0]) < 0).any():
+        raise ValueError("bands must be monotonically nondecreasing and have "
+                         "width > 0.")
+    if (bands[:-1, 1] > bands[1:, 0]).any():
+        raise ValueError("bands must not overlap.")
+    if (desired < 0).any():
+        raise ValueError("desired must be non-negative.")
+    if weight is None:
+        weight = cupy.ones(len(desired))
+    weight = cupy.asarray(weight).flatten()
+    if len(weight) != len(desired):
+        raise ValueError("weight must be the same size as the number of "
+                         "band pairs ({}).".format(len(bands)))
+    if (weight < 0).any():
+        raise ValueError("weight must be non-negative.")
+
+    # Set up the linear matrix equation to be solved, Qa = b
+
+    # We can express Q(k,n) = 0.5 Q1(k,n) + 0.5 Q2(k,n)
+    # where Q1(k,n)=q(k-n) and Q2(k,n)=q(k+n), i.e. a Toeplitz plus Hankel.
+
+    # We omit the factor of 0.5 above, instead adding it during coefficient
+    # calculation.
+
+    # We also omit the 1/π from both Q and b equations, as they cancel
+    # during solving.
+
+    # We have that:
+    #     q(n) = 1/π ∫W(ω)cos(nω)dω (over 0->π)
+    # Using our nomalization ω=πf and with a constant weight W over each
+    # interval f1->f2 we get:
+    #     q(n) = W∫cos(πnf)df (0->1) = Wf sin(πnf)/πnf
+    # integrated over each f1->f2 pair (i.e., value at f2 - value at f1).
+    n = cupy.arange(numtaps)[:, cupy.newaxis, cupy.newaxis]
+    q = cupy.dot(cupy.diff(cupy.sinc(bands * n) *
+                           bands, axis=2)[:, :, 0], weight)
+
+    # Now we assemble our sum of Toeplitz and Hankel
+    Q1 = toeplitz(q[:M+1])
+    Q2 = hankel(q[:M+1], q[M:])
+    Q = Q1 + Q2
+
+    # Now for b(n) we have that:
+    #     b(n) = 1/π ∫ W(ω)D(ω)cos(nω)dω (over 0->π)
+    # Using our normalization ω=πf and with a constant weight W over each
+    # interval and a linear term for D(ω) we get (over each f1->f2 interval):
+    #     b(n) = W ∫ (mf+c)cos(πnf)df
+    #          = f(mf+c)sin(πnf)/πnf + mf**2 cos(nπf)/(πnf)**2
+    # integrated over each f1->f2 pair (i.e., value at f2 - value at f1).
+    n = n[:M + 1]  # only need this many coefficients here
+    # Choose m and c such that we are at the start and end weights
+    m = (cupy.diff(desired, axis=1) / cupy.diff(bands, axis=1))
+    c = desired[:, [0]] - bands[:, [0]] * m
+    b = bands * (m*bands + c) * cupy.sinc(bands * n)
+    # Use L'Hospital's rule here for cos(nπf)/(πnf)**2 @ n=0
+    b[0] -= m * bands * bands / 2.
+    b[1:] += m * cupy.cos(n[1:] * cupy.pi * bands) / (cupy.pi * n[1:]) ** 2
+    b = cupy.diff(b, axis=2)[:, :, 0] @ weight
+
+    # Now we can solve the equation : XXX CuPy failure modes (?)
+    with cupyx.errstate(linalg="raise"):
+        try:
+            a = solve(Q, b)
+        except LinAlgError:
+            # in case Q is rank deficient
+            a = lstsq(Q, b, rcond=None)[0]
```

Comment:
Not sure if this is a right way of detecting an ill-conditioned linear algebraic system. scipy.signal version relies on `scipy.linalg.solve` emitting a LinAlgWarning , https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.solve.html

Would appreciate a pointer @takagi

Reply:
Looking at the `scipy.linalg.solve`, it appears that it uses separate steps for LU factorization and backsubstitution [1], so that it has a chance to call lapack's `*gecon` to estimate the condition number. CuPy linalg.solve, OTOH, calls straight into `*gesv` which does not return the condition number info [2]. 

So the procedure above is not entirely correct really. However, let's look at the tests. The procedure in this PR does not pass value-to-value comparisons between scipy and cupy, but does pass the test from scipy.signal [3] locally (currently commented out).

[1] https://github.com/scipy/scipy/blob/v1.10.1/scipy/linalg/_basic.py#L40-L265
[2] https://github.com/cupy/cupy/blob/v12.0.0/cupy/linalg/_solve.py#L15
[3] https://github.com/cupy/cupy/pull/7537/files#diff-15f8bbf4433eed91ce1cc13976ead22a49f85bc207b85c04a28a896a62b49960R93

Reply:
Sorry for dropping the ball. `cupyx.scipy.linalg.solve` doesn't raise here with expected ill-conditioned sytem?
https://github.com/cupy/cupy/blob/60fe053011b751129c5ccc7e6828f4f23c42763e/cupyx/lapack.py#L71

Reply:
Thanks for circling back to it @takagi ! No, `solve` does not raise the error here. Looking at https://docs.nvidia.com/cuda/cusolver/#cusolverdn-dense-lapack-function-reference, ISTM that `dev_info` reports the error condition only, similar to what CPU LAPACK does. Here the scipy.signal version explicitly checks the condition number via `gecon`.
Is there an analog of  https://netlib.org/lapack/explore-html/dd/d9a/group__double_g_ecomputational_ga188b8d30443d14b1a3f7f8331d87ae60.html on CUDA?

Reply:
I don't think cuSolver has `gecon` equivalent.

Reply:
Can we compute the condition number using the result of `getrf` here?
https://github.com/cupy/cupy/blob/60fe053011b751129c5ccc7e6828f4f23c42763e/cupyx/lapack.py#LL69C21-L69C21

Reply:
BTW, does the value-to-value comparison pass with some relaxed tolerance? Also, I think it's okay as far as the test from scipy.signal passes as it checks the condition the result should have.

Reply:
So this is a bit different, it turns out. 
There are two tests with rank-deficient LSQ problems. One indeed passes with a lower tolerance. The other runs into https://github.com/scipy/scipy/issues/18533: in a nutshell, the solution rather strongly depends on the LAPACK driver for the LSQ problem.
So the updated PR now duplicates the scipy tests and links to the scipy issue.